### ISSUE TITLE
MenuBundle disable all the knp menu extensions

### ISSUE THREAD

**Issue by efeencheung** [symfony-cmf/menu-bundle#249]
Menu Bundle disabled all the knp menu extensions when knp menu bundle regist after the cmf menu bundle，because the knp menu extension pass can not find the knp_menu.factory service

```php
class AddExtensionsPass implements CompilerPassInterface
{
    public function process(ContainerBuilder $container)
    {   
        if (!$container->hasDefinition('knp_menu.factory')) {
            return;
        }   

        $definition = $container->getDefinition('knp_menu.factory');

        foreach ($container->findTaggedServiceIds('knp_menu.factory_extension') as $id => $tags) {
            foreach ($tags as $tag) {
                $priority = isset($tag['priority']) ? $tag['priority'] : 0;
                $definition->addMethodCall('addExtension', array(new Reference($id), $priority));
            }   
        }   
    }   
}
```

**Comment by dbu** [comment]
i agree with you that the order of bundles must not be relevant for this.

but i am a bit confused to understand why this would be the case? i can't see the DI class checking for any knp menu services, and in compiler passes, all services from all bundles will be available.
what is the compiler pass you copied here?

**Comment by dbu** [comment]
ah, i think i got it. there is https://github.com/symfony-cmf/menu-bundle/blob/master/DependencyInjection/Compiler/DecorateMenuFactoryPass.php which replaces knp_menu.factory with cmf_menu.factory.quiet but sets up an alias. however, i guess the compiler pass of knp menu will then try to add the method calls to the alias which is not possible.

@WouterJ i checked but can't see anything about priorities of compiler passes. could we claim that this is a PassConfig::TYPE_OPTIMIZE so it always happens after the knp menu pass?

**Comment by efeencheung** [comment]
This is the knp menu extension source code. I just want to say the knp menu extension need the service knp_menu.factory when complie. But the DecorateMenuFactoryPass change the service name

**Comment by efeencheung** [comment]
I give a PR for fix it later

**Comment by efeencheung** [comment]
yeah, I suggest add the PassConfig::TYPE_BEFORE_REMOVING

**Comment by WouterJ** [comment]
I would rather fix this on the KnpMenuBundle side, by using `has()` and `findDefinition()` instead of `hasDefinition()` and `getDefinition()`. There will then by no problems anymore

**Comment by dbu** [comment]
ah, findDefinition will resolve the alias? nice. @efeencheung so can you do a PR to fix that on knp menu bundle side instead please?

i guess we should scan the cmf code for getDefinition then and replace that with findDefition as well.

**Comment by efeencheung** [comment]
This is a better solution! Thanks @WouterJ

**Comment by efeencheung** [comment]
I use 'has' instead of 'hasDefinition' will cause another problem, Do you do this now? @WouterJ  I think changing the DecorateMenuFactoryPass priorities is better for this bug.

**Comment by dbu** [comment]
@WouterJ the problem is that findDefinition finds our quiet factory, not the real factory. and the knp compiler pass assumes that the factory has a addExtension method, but that method is not in the interface.

i guess either we do #250 or we add that method to our quiet factory and fix the knp menu pass.

**Comment by dbu** [comment]
pull request on knp menu: https://github.com/KnpLabs/KnpMenuBundle/pull/317

**Comment by dbu** [comment]
and fix our part in #251.

the workaround until those are merged is to instantiate knp menu bundle before cmf menu bundle.
when moving to the decorator functionality, we can hopefully get rid of this workaround.
